File: packages/repo-gc-shared/src/heuristics/stringly-typed.ts
import { Finding, FindingKind, Severity, type FileData, type Thresholds } from '../types';

export function analyzeStringlyTyped(
  data: FileData,
  thresholds: Thresholds,
  idCounter: { value: number },
): Finding | null {
  const count = data.stringComparisonCount;
  const limit = thresholds.stringComparisonLimit;
  if (count <= limit) return null;

  let severity: Severity;
  if (count >= limit * 3) {
    severity = Severity.High;
  } else if (count >= limit * 2) {
    severity = Severity.Medium;
  } else {
    severity = Severity.Low;
  }

  return {
    id: `st-${String(idCounter.value++).padStart(3, '0')}`,
    kind: FindingKind.StringlyTyped,
    severity,
    confidence: 0.65,
    path: data.relativePath,
    summary: '',
    reasons: [],
    evidence: [
      `string_comparison_count: ${count}`,
      `limit: ${limit}`,
      `preview: ${data.relativePath}`,
    ],
    suggested_next_step: '',
  };
}

File: packages/repo-gc-python/python/repo_gc_python/heuristics/stringly_typed.py
"""Stringly-typed heuristic.

Detects magic string comparisons — string literals used in comparisons and
match patterns where enums or constants should be used instead.
"""

from ..cli import Threshold
from ..discovery import PythonFile
from ..parsing import FileInfo
from ..types import Finding, FindingKind, Severity


def analyze(
    pfile: PythonFile,
    info: FileInfo,
    threshold: Threshold,
    counter: int,
) -> Finding | None:
    limit = threshold.string_comparison_limit
    if info.string_comparison_count <= limit:
        return None

    if info.string_comparison_count >= limit * 3:
        severity = Severity.High
    elif info.string_comparison_count >= limit * 2:
        severity = Severity.Medium
    else:
        severity = Severity.Low

    return Finding(
        id=f"st-{counter:03d}",
        kind=FindingKind.StringlyTyped,
        severity=severity,
        confidence=0.65,
        path=str(pfile.relative_path),
        summary="",
        reasons=[],
        evidence=[
            f"string_comparison_count: {info.string_comparison_count}",
            f"limit: {limit}",
            f"preview: {pfile.relative_path}",
        ],
        suggested_next_step="",
        estimated_tokens=None,
    )
